fix kube-proxy latency metrics

aojea opened the pull request:
kube-proxy expose the metric network_programming_duration_seconds,
that is defined as the time it takes to program the network since
a a service or pod has changed. It uses an annotation on the endpoints
/endpointslices to calculate when the endpoint was created, however,
on restarts, kube-proxy process all the endpoints again, no matter
when those were generated, polluting the metrics.

To be safe, kube-proxy will estimate the latency only for those
endpoints that were generated after it started.

/kind bug
```release-note
fix a bug in kube-proxy latency metrics to calculate only the latency value for the endpoints that are created after it starts running. This is needed because all the endpoints objects are processed on restarts, independently when they were generated.
```

aojea commented:
/assign @smarterclayton

smarterclayton commented:
Looks correct.  The outcome I was seeing was prometheus reported a p99 latency of 300s - that was because about 5% of latency measurements were in the +Inf bucket, and 300 is the highest numbered bucket.  It is more accurate to report latency only when it is truly known - if necessary we could have a separate metric tracking "endpoints without any recorded change time" (which is 5% today as measured).

aojea commented:
/hold
the code in the cache checks that is zero, I need to understand this better

dcbw commented:
/triage accepted

JornShen commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -1620,6 +1620,12 @@ func (proxier *Proxier) syncProxyRules() {
+			// the EndpointsLastChangeTriggerTime annotation, so it is impossible
+			// to estimate the latency.
+			if lastChangeTriggerTime.IsZero() {
+				continue
```
add some warn log?
And Can you share your logs or metrics data of kube-proxy.

aojea replied:
I need to revisit it , this is not the problem ... :/

aojea committed:
kube-proxy network programming latency on restarts

kube-proxy expose the metric network_programming_duration_seconds,
that is defined as the time it takes to program the network since
a a service or pod has changed. It uses an annotation on the endpoints
/endpointslices to calculate when the endpoint was created, however,
on restarts, kube-proxy process all the endpoints again, no matter
when those were generated, polluting the metrics.

To be safe, kube-proxy will estimate the latency only for those
endpoints that were generated after it started.

aojea commented:
/hold cancel
/assign @wojtek-t @thockin

wojtek-t commented:
/lgtm
/approve

Thanks a lot for fixing this!

thockin commented on pkg/proxy/service_test.go:
```diff
@@ -511,12 +512,21 @@ type FakeProxier struct {
-		serviceChanges:   NewServiceChangeTracker(nil, ipFamily, nil, nil),
-		endpointsMap:     make(EndpointsMap),
-		endpointsChanges: NewEndpointChangeTracker(testHostname, nil, ipFamily, nil, false, nil),
+		serviceMap:     make(ServiceMap),
```
did you run gofmt on this?

aojea replied:
my editor does when it saves the file, it is wrong indented right?

aojea replied:
I run it manually and it keeps it this way, :shrug:

thockin replied:
OK.  Not a huge deal, maybe it changed in gofmt.

thockin commented:
Good catch

/lgtm
/approve

aojea commented:
this job failure is interesting, lots of test failed because pods can not be created in the same time frame, 20:52 - 20:55
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/100861/pull-kubernetes-e2e-kind/1379892041260994560
/retest

yangjunmyfm192085 commented:
/test pull-kubernetes-e2e-kind
